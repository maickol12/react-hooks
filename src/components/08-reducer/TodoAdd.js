import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo}) => {
    const [values, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmit = (e) =>{
        e.preventDefault();
        
        if(values.description.trim().length <= 1){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: values.description,
            done: false
        };
    

        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
           <h4>Agregar todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    className='form-control'
                    type='text'
                    value={ values.description }
                    onChange={ handleInputChange }
                    name='description'
                    placeholder='Aprender...'
                    autoComplete='off'></input>
                <button type='submit' className='btn btn-primary mt-1 btn-block'>
                    Agregar
                </button>
            </form>  
        </>
    )
}
