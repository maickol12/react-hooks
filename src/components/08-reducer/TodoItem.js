import React from 'react'

export const TodoItem = ({value,handleToggle,deleteTodo,indice}) => {
    const { id, done, desc } = value;
    return (
        <li key={id} className='list-group-item'>
            <p 
                className={`${done && 'complete'}`}
                onClick={() => handleToggle(id) }>
                {indice + 1}. {desc}
            </p>
            <button 
                onClick={ () => deleteTodo(id) }
                className='btn btn-danger'>
                Borrar
            </button>
        </li>
        
    )
}
