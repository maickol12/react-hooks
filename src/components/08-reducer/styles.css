body{
    padding: 70px;
}

p{
    margin: 0px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}