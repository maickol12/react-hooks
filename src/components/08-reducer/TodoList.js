import React from 'react';
import { TodoItem } from './TodoItem';

export const TodoList = ({todos,handleToggle,deleteTodo}) => {
    return (
        <div className='col-5'>
            <ul className='list-group list-group-flush'>
            { todos.map((value,indice) => (
                       <TodoItem
                            key={`todoItem${value.id}`}
                            value={value}
                            handleToggle={handleToggle}
                            deleteTodo={deleteTodo}
                            indice={indice} /> 
                    )
                    ) 
                }
            </ul>
        </div>
    )
}
