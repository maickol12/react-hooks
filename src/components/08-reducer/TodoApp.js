import React, { useReducer, useState, useEffect } from 'react'
import { useForm } from '../../hooks/useForm';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || []
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprendiendo react',
    //     done: false
    // }];
}


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos])



    const deleteTodo = (id) => {
        
        const action = {
            type:'delete',
            payload: id
        }

        dispatch(action);

    }

    const handleToggle = (id) => {
        dispatch({type:'toggle',payload:id});
    }
    
    const handleAddTodo = (newTodo) => {
        const action = {
            type:'add',
            payload: newTodo
        }
        dispatch(action);
    }

    return (
        <div>
            <h1>TodoAPP ({ todos.length})</h1>
            <hr />
            <div className='row'>
                <TodoList 
                    todos={todos}
                    handleToggle={handleToggle}
                    deleteTodo={deleteTodo} />
                <div className='col-7'>
                   <TodoAdd handleAddTodo={handleAddTodo} />
                </div>

            </div>
        </div>
    )
}
