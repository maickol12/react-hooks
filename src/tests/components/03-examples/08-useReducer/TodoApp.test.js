import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import { TodoApp } from "../../../../components/08-reducer/TodoApp";
import { demosTodos } from "../../../fixtures/demosTodos";

describe('Pruebas en <TodoApp />', () => {
    const wrapper = shallow(<TodoApp /> )
   test('debe de mostrarse correctamente ', () => {
       expect( wrapper ).toMatchSnapshot();
   });
   
   test('debe de agregar un TODO ', () => {
        const wrapper = mount( <TodoApp />  );
        act(() =>{
            wrapper.find('TodoAdd').prop('handleAddTodo')( demosTodos[0] );
            wrapper.find('TodoAdd').prop('handleAddTodo')( demosTodos[1] );
        });


        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp ( 2 )');
        expect( localStorage.setItem ).toHaveBeenCalledTimes( 2 );
    });
    test('debe de elimonar un TODO ', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demosTodos[0]);
        wrapper.find('TodoAdd').prop('handleAddTodo')(demosTodos[0].id);
        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp ( 1 )');
    });
    
});
