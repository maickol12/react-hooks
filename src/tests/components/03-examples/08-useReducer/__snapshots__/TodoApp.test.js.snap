// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe de mostrarse correctamente  1`] = `
<div>
  <h1>
    TodoAPP (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <TodoList
      deleteTodo={[Function]}
      handleToggle={[Function]}
      todos={Array []}
    />
    <div
      className="col-7"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
